### PR TITLE
updateApis shouldn't fail if it's unable to fetch an asset

### PR DESCRIPTION
**Pull request by gkhunger-sf**
- Display a message to manually download a RAML if exhcnage-downloader couldn't get it
- Increase the code coverage threshold to 85%
- Refactor exchange-downloader tests to use stubs for helper functions

### CODE REVIEW

File: packages/exchange-connector/src/exchangeDownloader.ts [hunk 1]
```diff
@@ -12,21 +12,24 @@ import fetch, { Response } from "node-fetch";
...
 
 import { RestApi, FileInfo, Categories } from "./exchangeTypes";
+import { sdkLogger } from "../../core/src";
 
 const DEFAULT_DOWNLOAD_FOLDER = "download";
 const ANYPOINT_BASE_URI = "https://anypoint.mulesoft.com/exchange/api/v2";
+const ANYPOINT_BASE_URI_WITHOUT_VERSION =
+  "https://anypoint.mulesoft.com/exchange";
 
 export function downloadRestApi(
```
Comment (bredmond-sf) [inline packages/exchange-connector/src/exchangeDownloader.ts:22]: Just a suggestion, but this could be simplified a bit...
```
export function async downloadRestApi(
  restApi: RestApi,
  destinationFolder: string = DEFAULT_DOWNLOAD_FOLDER
): Promise<void | Response> {

    if (!restApi.fatRaml?.createdDate) {
      sdkLogger.warn(
        `Failed to download '${restApi.name}' RAML as Fat RAML download information is missing.`,
        `Please download it manually from ${ANYPOINT_BASE_URI_WITHOUT_VERSION}/${restApi.assetId} and update the relevant details in apis/api-config.json`
      );
      return;
    }

    ensureDirSync(destinationFolder);
    const zipFilePath = path.join(destinationFolder, `${restApi.assetId}.zip`);
    const response = await fetch(restApi.fatRaml.externalLink);
    writeFileSync(zipFilePath, response.body);
    return response;
}
```

File: packages/exchange-connector/src/exchangeDownloader.ts [hunk 2]
```diff
@@ -164,24 +170,30 @@ export function getVersionByDeployment(
   restApi: RestApi,
   deployment: RegExp
 ): Promise<string> {
-  return getAsset(accessToken, `${restApi.groupId}/${restApi.assetId}`).then(
-    asset => {
+  return exports
```
Comment (bredmond-sf) [inline packages/exchange-connector/src/exchangeDownloader.ts:173]: same question

File: packages/exchange-connector/src/exchangeDownloader.ts [hunk 3]
```diff
@@ -54,7 +57,7 @@ export function downloadRestApis(
   const promises: Promise<any>[] = [];
 
   restApi.forEach((api: RestApi) => {
-    promises.push(downloadRestApi(api, destinationFolder));
+    promises.push(exports.downloadRestApi(api, destinationFolder));
```
[the remark this answers is not in the document]
Comment (gkhunger-sf) [reply packages/exchange-connector/src/exchangeDownloader.ts:60]: to help with stubbing

File: packages/exchange-connector/gulpfile.ts [hunk 4]
```diff
@@ -66,7 +87,7 @@ function downloadRamlFromExchange(): Promise<void> {
   return search().then(apis => {
     return downloadRestApis(apis, config.inputDir)
       .then(folder => {
-        console.log(`Setting config.inputDir to '${folder}'`);
+        sdkLogger.warn(`Setting config.inputDir to '${folder}'`);
```
[the remark this answers is not in the document]
Comment (gkhunger-sf) [reply packages/exchange-connector/gulpfile.ts:90]: it isn't 😛 
will change to info

File: packages/exchange-connector/gulpfile.ts [hunk 5]
```diff
@@ -19,12 +19,33 @@ import {
   getSpecificApi,
   groupByCategory
 } from "./src";
+import { sdkLogger } from "../core";
```
Comment (monkey-jeff) [inline packages/exchange-connector/gulpfile.ts:22]: [ISSUE] Should not have relative dependencies across packages.

File: packages/exchange-connector/gulpfile.ts [hunk 6]
```diff
@@ -19,12 +19,33 @@ import {
...
 
+/**
+ * Remove all the information that could be invalid so that the SDK is not
+ * generated with incorrect information accidentally. Correct information
+ * needs to be filled manually in 'apis/api-config.json' before the SDK can be
+ * generated.
+ *
+ * @param api The API with incorrect information
+ */
+function removeInvalidInformation(api: RestApi) {
```
Comment (monkey-jeff) [inline packages/exchange-connector/gulpfile.ts:37]: The method should not be in the gulpfile.  Gulpfiles should be as thin as possible.

File: packages/exchange-connector/src/exchangeDownloader.ts [hunk 7]
```diff
@@ -12,21 +12,24 @@ import fetch, { Response } from "node-fetch";
 import path from "path";
 
 import { RestApi, FileInfo, Categories } from "./exchangeTypes";
+import { sdkLogger } from "../../core/src";
```
Comment (monkey-jeff) [inline packages/exchange-connector/src/exchangeDownloader.ts:15]: Same as above.  Don't use relative imports across packages.

File: packages/exchange-connector/src/exchangeDownloader.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (monkey-jeff) [reply packages/exchange-connector/src/exchangeDownloader.ts:60]: I have to say...I don't love this but will accept it if we don't have a better way.

File: packages/exchange-connector/src/exchangeDownloader.ts (hunk 1 above)
Comment (monkey-jeff) [reply packages/exchange-connector/src/exchangeDownloader.ts:22]: I like this proposal

File: packages/exchange-connector/src/exchangeDownloader.ts (hunk 2 above)
Comment (monkey-jeff) [reply packages/exchange-connector/src/exchangeDownloader.ts:173]: I like this even less here.

File: packages/exchange-connector/test/exchangeDownloader.test.ts [hunk 8]
```diff
@@ -67,84 +68,76 @@ describe("Test downloadRestApi", () => {
...
-      "Fat RAML download information for test-api is missing"
-    );
+    return expect(downloadRestApi(api)).to.eventually.be.undefined;
   });
 });
 
-describe("Test downloadRestApis", () => {
-  afterEach(nock.cleanAll);
+describe("downloadRestApis", () => {
+  beforeEach(sinon.restore);
```
Comment (monkey-jeff) [inline packages/exchange-connector/test/exchangeDownloader.test.ts:96]: Why did you switch from nock to sinon here?

File: packages/exchange-connector/test/exchangeDownloader.test.ts [hunk 9]
```diff
@@ -153,30 +146,28 @@ describe("Test searchExchange", () => {
...
 
     return searchExchange("AUTH_TOKEN", "searchString").then(res => {
-      expect(res).to.be.deep.equal(searchAssetApiResultObject);
+      expect(res).to.deep.equal(searchAssetApiResultObject);
     });
   });
 });
 
-describe("Test getSpecificApi", () => {
-  afterEach(nock.cleanAll);
```
Comment (monkey-jeff) [inline packages/exchange-connector/test/exchangeDownloader.test.ts:162]: Same question

File: packages/generator/src/parser.ts [hunk 10]
```diff
@@ -15,6 +15,7 @@ import amf from "amf-client-js";
 import path from "path";
 import _ from "lodash";
 import { RestApi } from "@commerce-apps/exchange-connector";
+import { sdkLogger } from "../../core/src";
```
Comment (monkey-jeff) [inline packages/generator/src/parser.ts:18]: Same as before.  Do not have cross package relative dependencies.

File: packages/generator/src/parser.ts [hunk 11]
```diff
@@ -97,6 +98,10 @@ export function processApiFamily(
   const promises = [];
   const ramlFileFromFamily = apiFamilyConfig[apiFamily];
   _.map(ramlFileFromFamily, (apiMeta: RestApi) => {
+    if(!apiMeta.fatRaml.createdDate) {
```
Comment (monkey-jeff) [inline packages/generator/src/parser.ts:101]: Why are we only checking for createdDate?

File: package.json [hunk 12]
```diff
@@ -3,7 +3,7 @@
   "private": true,
   "scripts": {
     "postinstall": "lerna bootstrap --hoist",
-    "updateApis": "lerna run updateApis",
+    "updateApis": "lerna run updateApis --stream",
```
Comment (monkey-jeff) [inline package.json:6]: Why this change here?

File: packages/exchange-connector/src/exchangeDownloader.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gkhunger-sf) [reply packages/exchange-connector/src/exchangeDownloader.ts:60]: I know it's not ideal but it's either this or moving the helper functions to a separate file.

File: packages/exchange-connector/test/exchangeDownloader.test.ts (hunk 8 above)
Comment (gkhunger-sf) [reply packages/exchange-connector/test/exchangeDownloader.test.ts:96]: To stub the helper functions, sinon was only mocking the calls made by the helper functions. I believe the tests shouldn't be invoking the helper functions unless they are the ones under test.

File: packages/generator/src/parser.ts (hunk 11 above)
Comment (gkhunger-sf) [reply packages/generator/src/parser.ts:101]: createdDate is one of the fields that we reset when we are unable to get an asset from exchange. Since it is one of the mandatory fields, we shouldn't go forward if this is not set.

File: package.json (hunk 12 above)
Comment (gkhunger-sf) [reply package.json:6]: lerna doesn't show logs from the child processes by default, it only shows the end result

File: packages/exchange-connector/src/exchangeDownloader.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (unandyala) [reply packages/exchange-connector/src/exchangeDownloader.ts:60]: I don't think we want this approach, I documented the details in the renderer tests
```javascript
/**
 * Tests all the functions that are invoked while rendering templates.
 *
 * Note: Ideally we want a independent test for each function. As of today (01/30/2020) this is not possible with Sinon when the functions are in the same module.
 * Sinon works by overwriting a reference to a function on an object, making it point at something else. It can't replace references to the same function in other contexts.
 *
 * Example: We can't mock renderApiFamily function while testing renderTemplates function. To make it work renderTemplates should invoke
 * renderApiFamily as "module.exports.renderApiFamily(..)" which is not desired.
 */
```
